### ISSUE TITLE
Support pure pytorch implementation for memory_efficient_attention

### ISSUE THREAD

**Issue by hoangmit** [facebookresearch/xformers#536]
# 🚀 Feature

I found that `memory_efficient_attention` op does not support pure pytorch implementation (e.g. without using device specific op, or library or cython). The current implementations fail to dispatch on CPU device.

In the spirit of torch 2.0, we should have a python implementation for most op.

Something similar to: https://github.com/lucidrains/memory-efficient-attention-pytorch

Pure pytorch implementation is useful for testing, benchmarking, and generic support purposes.

**Comment by blefaudeux** [comment]
The algorithm from Rabe that you linked to is no the full Flash Attention, it only covers the initial softmax(QKt) but flash attention does the V dot product at the same time and it makes it a lot more effective. This algorithm cannot (to the best of my knowledge) be written using pytorch's primitives, you can bet that it would be used everywhere already if that was the case. Feel free to write a PR to prove me wrong though :)

**Comment by danthe3rd** [comment]
I think the point is to have a fall-back for CPU (and unsupported devices) that is memory-efficient - even if it's much slower than pytorch

**Comment by blefaudeux** [comment]
ah ok, that's not how I understood the title and issue. In that case I'm not sure that there's much to be done beyond slicing per row on the attention matrix, is that what you meant @hoangmit ?
